internal/policy/policy.go

```diff
@@ -230,6 +260,50 @@ func (s *State) FindAuthorizedSigningKeyIDs(ctx context.Context, roleName string
 	return entry.KeyIDs, nil
 }
 
+func (s *State) FindPublicKeysForPath(ctx context.Context, path string) ([]*tuf.Key, error) {
+	if err := s.Verify(ctx); err != nil {
+		return nil, err
+	}
+
+	targetsMetadata, err := s.GetTargetsMetadata(TargetsRoleName)
+	if err != nil {
+		return nil, err
+	}
+
+	allPublicKeys := targetsMetadata.Delegations.Keys
+	delegationsQueue := targetsMetadata.Delegations.Roles
+
+	trustedKeys := []*tuf.Key{}
+	for {
+		if len(delegationsQueue) <= 1 {
+			return trustedKeys, nil
+		}
+
+		delegation := delegationsQueue[0]
+		delegationsQueue = delegationsQueue[1:]
+
+		if delegation.Matches(path) {
+			for _, keyID := range delegation.KeyIDs {
+				key := allPublicKeys[keyID]
+				trustedKeys = append(trustedKeys, key)
+			}
+
+			// TODO: clarify logic for delegation.Terminating
+
+			if s.HasTargetsRole(delegation.Name) {
+				delegatedMetadata, err := s.GetTargetsMetadata(delegation.Name)
+				if err != nil {
+					return nil, err
+				}
+				for keyID, key := range delegatedMetadata.Delegations.Keys {
+					allPublicKeys[keyID] = key
+				}
+				delegationsQueue = append(delegationsQueue[:len(delegationsQueue)-1], delegatedMetadata.Delegations.Roles...)
```

Comment:
This is taking all but the last element of the queue and then adding the new target roles.

It's not immediately obvious to me why this is correct (e.g. why are we ignoring the last element) - I'd add comments explaining this in more detail.